### PR TITLE
Distributed w2v fixes

### PR DESCRIPTION
**Pull request by MissingNO57**
W2V dataloader end of file bug fix
W2V client data splitting fixed
Client learning rate decay fixed
Removed leftover code
Added client's train and test loss member variables

### CODE REVIEW

File: libs/ml/examples/word2vec_distributed_learning/main.cpp [hunk 1]
```diff
@@ -55,9 +56,19 @@ std::vector<std::string> SplitTrainingData(std::string const &train_file,
...
+      // Last client gets the rest
+      pos = input_data.size();
+    }
+    else
+    {
+      // find next instance of space character
+      pos = input_data.find(" ", pos, 1);
+    }
+
+    client_data.push_back(input_data.substr(old_pos, pos - old_pos));
```
Comment (evsmithx) [inline libs/ml/examples/word2vec_distributed_learning/main.cpp:61]: The if-else above is unnecessary now that substr uses (pos - old_pos). If the length is longer than the input string then it just goes to the end of the string.

**Review by evsmithx — DISMISSED** [review]
LGTM

File: libs/ml/include/ml/dataloaders/word2vec_loaders/sgns_w2v_dataloader.hpp [hunk 2]
```diff
@@ -396,7 +395,7 @@ void GraphW2VLoader<T>::BufferNextSamples()
             window_size_;  // the current word start from position that allows full context window
         current_sentence_++;
       }
-      if (IsDone())
+      if (IsDone() || current_sentence_ >= data_.size())
```
Comment (evsmithx) [inline libs/ml/include/ml/dataloaders/word2vec_loaders/sgns_w2v_dataloader.hpp:397]: If current_sentence_ >= data_.size() won't IsDone() be true anyway? I think the second condition here is unnecessary.

Comment (MissingNO57) [reply libs/ml/include/ml/dataloaders/word2vec_loaders/sgns_w2v_dataloader.hpp:397]: IsDone is checking for data in buffer and when there are data it continues, but we can be in next non-existing sentence and crash. This was the only way I found to pass W2V tests and not be crashing on distributed W2V training.

**Review by baykaner — APPROVED** [review]
